### PR TITLE
Dev/saars/config perf counter

### PR DESCRIPTION
**Pull request by xiaomi7732**
Address #217.

* Refactor the telemetries to separate the simple perf counter telemetry initializer with the primary k8s telemetry initializer
* Honor the user options to decide to include the simple perf counter telemetry initializer or not.
* Readme is updated with regarding the new option.
* CPU / Memory metrics have been moved to metrics of a telemetry entry from customer dimensions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -93,6 +93,7 @@ Customize configurations are supported starting with version 1.0.2 of the Applic
     | Key                   | Value/Types | Default Value | Description                                                                                  |
     | --------------------- | ----------- | ------------- | -------------------------------------------------------------------------------------------- |
     | InitializationTimeout | TimeSpan    | 00:02:00      | Maximum time to wait for spinning up the container. Accepted format: [d.]hh:mm:ss[.fffffff]. |
+    | DisableCounters       | Boolean     | false         | Sets to true to avoid adding performance counter telemetry initializer.                      |
```
Comment (karolz-ms) [inline README.md:96]: Consider `DisablePerformanceCounters` for extra clarity

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            if (telemetry is ISupportMetrics metricsTelemetry)
+            {
+                // Write CPU metrics
+                TimeSpan endCPUTime = GetCPUTime();
+                if (_lastCPUSample != default)
+                {
+                    if (_cpuWatch.ElapsedMilliseconds > 0)
+                    {
+                        int processorCount = Environment.ProcessorCount;
```
Comment (karolz-ms) [inline src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:27]: This does not seem to be used in the CPU time calculation ☺️ Also the value could probably be cached

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        private void SetMetrics(ISupportMetrics telemetry, string key, double value)
+        {
+            if (!telemetry.Metrics.ContainsKey(key))
+            {
+                telemetry.Metrics.Add(key, value);
+            }
+        }
+
+        private TimeSpan GetCPUTime() => GetFromCurrentProcess<TimeSpan>(p => p.TotalProcessorTime);
```
Comment (karolz-ms) [inline src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:57]: Creating and disposing of a `Process` object is somewhat expensive. Consider caching it.

**Review by karolz-ms — CHANGES_REQUESTED** [review]

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs (hunk 2 above)
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:27]: Good catch. It had been missed...

File: README.md (hunk 1 above)
Comment (xiaomi7732) [reply README.md]: Point taken. Updated.

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs (hunk 2 above)
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs]: Fixed in the latest code.

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs (hunk 3 above)
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:62]: Done.

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/KubernetesTelemetryInitializer.cs [hunk 4]
```diff
@@ -229,6 +193,6 @@ private static void SetCustomDimension(ISupportProperties telemetry, string key,
 
         internal bool _isK8sQueryTimeout = false;
         private bool _isK8sQueryTimeoutReported = false;
-        private static readonly ApplicationInsightsKubernetesDiagnosticSource _logger = ApplicationInsightsKubernetesDiagnosticSource.Instance;
+        protected static readonly ApplicationInsightsKubernetesDiagnosticSource _logger = ApplicationInsightsKubernetesDiagnosticSource.Instance;
```
Comment (xiaomi7732) [inline src/ApplicationInsights.Kubernetes/TelemetryInitializers/KubernetesTelemetryInitializer.cs]: Change back.

Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/TelemetryInitializers/KubernetesTelemetryInitializer.cs]: Done

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        private long GetMemory() => GetFromCurrentProcess<long>(p => p.VirtualMemorySize64);
+
+        private T GetFromCurrentProcess<T>(Func<Process, T> valueFactory)
+        {
+            // In case of concurrent execution, there is a chance for _currentProcess to be checked null more than one time,
+            // Take the hit to assign it multiple times to avoid locks.
+            if (_currnetProcess == null)
```
Comment (karolz-ms) [inline src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:69]: ... or you can use `Lazy<T>` (just a thought)

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace Microsoft.ApplicationInsights.Kubernetes
+{
+    internal sealed class SimplePerformanceCounterTelemetryInitializer : ITelemetryInitializer
+    {
+        private TimeSpan _lastCPUSample;
+        private Stopwatch _cpuWatch = new Stopwatch();
+
+        // Cache current process for performance trade off. 
+        // Refer to https://stackoverflow.com/questions/28023490/should-one-call-dispose-for-process-getcurrentprocess
```
Comment (karolz-ms) [inline src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:18]: I think it is OK to not bother with disposing the Process object because `SimplePerformanceCounterTelemetryInitializer` is used as a **singleton** (one per Application Insights telemetry processor chain). That is a better justification IMO than relying on some Process implementation details.

**Review by karolz-ms — APPROVED** [review]
LGTM, just a couple minor comments.

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs (hunk 5 above)
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:69]: Yes. Thanks for bring this up. Funny that I started with Lazy<T> and decide not to go with it just to make it even easier. The part that I am not 100% comfortable with Lazy<T> in this case is that Process is actually a IDisposable. I'll resolve the conversation for now but good point to think about.

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs (hunk 6 above)
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:18]: Thanks for bring this up. I am sorry but I only partially agree :-).

I agree rely on the callee's implementation details is not ideal.

However, IMHO, it is as not ideal as rely on caller's implementation - there is NO constraint saying that `SimplePerformanceCounterTelemetryInitializer` or any `ITelemetryInitializer` have to be used as a singleton either.

That said, this crossed my mind, I just think it too heavy to implement for now. Post it here totally for discussion:
Basically, abstract out the current process part.
To create a singleton `CurrentProcessorInfoProvider`, that provides a method to get info from the current process. Implementation wise, current process object will be lazy load and cached, and it could implement IDisposable.

Let me know what do you think. :-D

Comment (karolz-ms) [reply src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:18]: I like your proposal a lot, with just one caveat: as far as I can tell, the property getters of the `Process` class are not thread-safe, so the provider code will probably have to ensure that only one thread is accessing the `Process` instance at any given time

Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:18]: Task for another day. TBH, at this moment, I am more willing to rule out the SimplePerformanceCounterTelemetryInitializer. Here are my reasons:
1. **It is NOT k8s specific, and doesn't have to be in this repository;** (shall I say it shouldn't be here?)
1. .NET Core 3.x provides performance counters, I think it will be way more reliable to use that as the implementation to get metrics like CPU/Memory usage.

Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/TelemetryInitializers/SimplePerformanceCounterTelemetryInitializer.cs:18]: All that said, thank you so much for the review!
